### PR TITLE
implementation, before testing

### PR DESCRIPTION
**Pull request by lilachmaliniak**
seamless failover implementation

### CODE REVIEW

File: lib/entry/crossDCsStreamPersistence.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+//   2.2) save playlist. ---> using this.playlistGenerator
+//   2.3) remove from list
+// 3) add sequence to restore _entriesToProcess, when LiveController starts!!!
+
+class CrossDCsStreamPersistence {
+
+    constructor() {
+        let endListPendingEntriesFileName = "endListPending.json";
+        let persistentEntriesMinimalDataFileName = "persistentEntries.json"
+        let localFolder = config.get('localFolderPath') || '.';
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: localFolderPath?

File: lib/entry/crossDCsStreamPersistence.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+//---------------------------------------------------------------
+// Todo: Lilach: PLAT-5023
+// 1) check if any DC provides the entryId:
+//  1.1) in case no DC provides the entry: update the time tag +
+//  1.2) otherwise,
+//      1.2.1) start timer to check if entry is provided
+//      1.2.2) rename playlist.json
+//--------------------------------------------------------------
+CrossDCsStreamPersistence.prototype.streamPersistenceCheck = function (entryId) {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why isn't this written in ES6? (inside the class )?

File: lib/entry/crossDCsStreamPersistence.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+    var deferred = Q.defer();
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why do you use defer? it's obselete

File: lib/entry/crossDCsStreamPersistence.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        if (now - this._entriesData[entryId].timestamp >= this.checkStreamIntervalMSec) {
+            this.streamPersistenceCheck(entryId);
+        }
+    });
+
+    return this.flushEntryList(this._entriesToProcess);
+
+};
+
+CrossDCsStreamPersistence.prototype.backwardCompatibleInit = function() {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why do we need this? we don't care about the transition ... it doesn't worth even 5 seconds...

File: lib/playlistGenerator/PlaylistGenerator.js [hunk 5]
```diff
@@ -267,7 +275,8 @@ var doPersist =  function() {
 };
 
 // save playlist
-var persistPlayList = function persistPlayList(force){
+// todo: if stream ended set presentationEndTime to now - playWindow
+var persistPlayList = function persistPlayList(force, ended = false){
```
Comment (kalturaguy) [inline lib/playlistGenerator/PlaylistGenerator.js]: it's a bit combersome... i think it'd be better if you expose an interface for the playlist generator that you will setup if it ended or now

File: lib/playlistGenerator/PlaylistGenerator.js [hunk 6]
```diff
@@ -374,10 +384,12 @@ PlaylistGenerator.prototype.stop = function stop (){
     }
     if(isRunning.call(that)){
         that.startPromise = null;
+        that.playlistImp.expirationTime = Data.now;
```
Comment (kalturaguy) [inline lib/playlistGenerator/PlaylistGenerator.js]: i'd rather you update the expirationTime every time you update the playlist
so if the process crashes/ machine killed it will automtecly be expired...

File: lib/playlistGenerator/PlaylistGenerator.js [hunk 7]
```diff
@@ -501,7 +513,31 @@ PlaylistGenerator.prototype.reset = function() {
...
+}
 
+// update presentationEndTime and save playlist
+PlaylistGenerator.prototype.persistStopped = function() {
+    return persistPlayList.call(this, true);
+}
 
+// update presentationEndTime and expireTime and save to disk
+PlaylistGenerator.prototype.restoreStopped = function() {
+    this.start();
```
Comment (kalturaguy) [inline lib/playlistGenerator/PlaylistGenerator.js]: no need for this , since you will update presentationEndTime and expirationTime when updating the playlist

File: lib/entry/crossDCsStreamPersistence.js [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+    return Q.fcall(function () {
+        if (that._entriesToProcess[entryId]) {
+            logger.info("[%s] entry is already in stream persistence list, ignoring request", entryId);
+            // todo: replace the playlist obj (consult with Guy)
+            // return Q.resolve(); required ???
+            return true;
+        } else {
+            // keep the playlistGenerator obj
+            this._entriesData[entryId] = { playlist: playlistGenerator, timestamp: new Date() };
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why not keep the only the playlistgenerator? the timestamp is already kept inside the playlistgenerate, just expose it

File: lib/entry/crossDCsStreamPersistence.js (hunk 1 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: backward compatibility... didn't right this code.

Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: after discussion with Guy, the code was removed, as decided.

File: lib/entry/crossDCsStreamPersistence.js (hunk 2 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: moved code to ES6 class methods.

File: lib/entry/crossDCsStreamPersistence.js (hunk 3 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: removed.

File: lib/entry/crossDCsStreamPersistence.js (hunk 4 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: removed

File: lib/playlistGenerator/PlaylistGenerator.js (hunk 5 above)
Comment (lilachmaliniak) [inline lib/playlistGenerator/PlaylistGenerator.js]: code changed. the playlist.json is read and then updated and written to disk directly. Changes in PlaylistGeneration reverted.

File: lib/playlistGenerator/PlaylistGenerator.js (hunk 6 above)
Comment (lilachmaliniak) [inline lib/playlistGenerator/PlaylistGenerator.js]: done.

File: lib/playlistGenerator/PlaylistGenerator.js (hunk 7 above)
Comment (lilachmaliniak) [inline lib/playlistGenerator/PlaylistGenerator.js]: code changed. this method no longer exist.

File: lib/entry/crossDCsStreamPersistence.js (hunk 8 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: code updated as mentioned above. PlaylistGenerator obj is not kept anymore. The time kept is the timestamp of last check if entry is provided by any DC.

File: packager/config/nginx.conf.template [hunk 9]
```diff
@@ -49,8 +49,8 @@ http {
         vod_align_segments_to_key_frames on;
 	    vod_response_cache response_cache 128m 30;
 	    vod_max_mapping_response_size 1m;
-        vod_live_mapping_cache live_mapping_cache 128m 2;
-        vod_live_response_cache live_response_cache 128m 2;
+        #vod_live_mapping_cache live_mapping_cache 128m 2;
```
Comment (kalturaguy) [inline packager/config/nginx.conf.template]: i think you got this mixed up with other bracnh

File: lib/playlistGenerator/PlaylistGenerator.js [hunk 10]
```diff
@@ -233,9 +233,9 @@ var doPersist =  function() {
     that.emit('beforePlaylistChanged', that);
 
     that.isDirty = false;
-    if(that.playlistImp.inner.liveWindowDuration) {
-        that.playlistImp.inner.presentationEndTime = that.playlistImp.getCurrentTime() + that.playlistImp.inner.liveWindowDuration;
-    }
+    that.playlistImp.inner.presentationEndTime = that.playlistImp.getCurrentTime() + that.playlistImp.inner.liveWindowDuration;
+    that.playlistImp.inner.expirationTime = that.playlistImp.inner.presentationEndTime;
```
Comment (kalturaguy) [inline lib/playlistGenerator/PlaylistGenerator.js]: there should be two values:
1. when to return 404
2. when to return end-list

those are different values, in your code it's the same

File: lib/entry/crossDCsStreamPersistence.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+            })
+            .then((isLive) => {
+                entryIsLive = isLive;
+                return qio.read(this._playlistsData[entryId].path);
+            })
+            .then((content) => {
+                let playlist = content;
+                this._playlistsData[entryId].lastProcessedTime = now;
+                if (newEntry) {
+                    logger.debug(`====> [streamPersistenceCheck] [${entryId}] STOPPED STREAMING expiration time set to [${nowStr}]`);
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: the logger should be consist with all our log (entry id in the begining)
look at the log file

File: lib/entry/crossDCsStreamPersistence.js [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+        logger.debug(`====> [addEntry] adding [${entryId}] to persistence list`);
+
+        if (this._entriesToProcess[entryId]) {
+            logger.debug("====> [addEntry][%s] entry is already in stream persistence list, ignoring request", entryId);
+            return Q.resolve();
+        } else {
+            this._entriesToProcess[entryId] = entryId;
+            this._playlistsData[entryId] = {
+                path: path.join(persistenceFormat.getEntryBasePath(entryId), persistenceFormat.getMasterManifestName()),
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why keep the path? entry id is enough, you can always calcualte it

File: lib/entry/crossDCsStreamPersistence.js [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+
+                    /*         // todo: rethink flash operation location and timing...
+                     return this.flushEntryList(this._entriesToProcess)
+                     .catch((err)=> {
+                     logger.error(`error restoring stopped entries list, error: [${ErrorUtils.error2string(err)}]`);
+                     return Q.resolve(true);
+                     });*/
+                })
+                .catch((err) => {
+                    logger.info(`====> [init] error restoring stopped entries list, error: [${ErrorUtils.error2string(err)}]`);
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: entry id is missing

File: lib/entry/crossDCsStreamPersistence.js [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+                });
+        }
+
+    }
+
+    removeEntry(entryId) {
+
+        logger.debug(`====> [removeEntry] removing [${entryId}] from persistence list`);
+        if (this._entriesToProcess.hasOwnProperty(entryId)) {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: use .has

File: lib/entry/crossDCsStreamPersistence.js [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+    addEntry(entryId) {
+
+        logger.debug(`====> [addEntry] adding [${entryId}] to persistence list`);
+
+        if (this._entriesToProcess[entryId]) {
+            logger.debug("====> [addEntry][%s] entry is already in stream persistence list, ignoring request", entryId);
+            return Q.resolve();
+        } else {
+            this._entriesToProcess[entryId] = entryId;
+            this._playlistsData[entryId] = {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why keep extra data structure? it's extra overhead, you just need to keep 
_entriesToProcess the same as we did in 0.15 brach

File: common/config/config.json.template [hunk 16]
```diff
@@ -24,15 +24,45 @@
   "logFileName" : "~/dvr/kLiveController.log",
   "logTimeZoneOffset" : 300,
   "monitorPort": 4390,
+  "persistenceCheckIntervalMillisec": 20000,
+  "stoppedEntriesPath": "~/kLivePersistence",
   "regressionAdapter": {
```
Comment (kalturaguy) [inline common/config/config.json.template]: this doesn't belong to this PR

File: common/config/config.json.template [hunk 17]
```diff
@@ -24,15 +24,45 @@
...
+             }
+        }]
   },
   "diagnostics": {
       "minChunkDurationInSec": 1,
       "maxChunkDurationInSec": 60,
       "errorsWindowSizeInMin" : 5,
       "ptsMaxMisalignmentAllowedInMSec" : 5000
   },
+  "preserveOriginalHLS": {
```
Comment (kalturaguy) [inline common/config/config.json.template]: this doesn't belong to this PR

File: lib/Controller.js [hunk 18]
```diff
@@ -70,8 +72,17 @@ module.exports = (function(){
...
 
+        this.crossDCsStreamPersistence.removeEntry(e.entryId);
+
         return liveEntry.start()
-            .then(function(){
+            .then(function () {
+                return liveEntry;
+            })
+            .then((liveEntry) => {
+                that.crossDCsStreamPersistence.flushEntryList()
```
Comment (kalturaguy) [inline lib/Controller.js]: why do we call flushEntry list here? the remove Entry that you call few lines ago should do it
the logic of the crossDCsStreamPersistence should not be handled from ourside

File: lib/Controller.js [hunk 19]
```diff
@@ -191,6 +202,9 @@ module.exports = (function(){
             else {
                 logger.info("[%s] Entry is no longer live, stopping it", noLongerLiveEntry);
                 return liveEntry.stop()
+                    .then(() => {
+                        return that.crossDCsStreamPersistence.addEntry(noLongerLiveEntry);
```
Comment (kalturaguy) [inline lib/Controller.js]: we need to be carfull here, since we don't want exception from addEntry to block our flow,
remove the return

File: lib/entry/crossDCsStreamPersistence.js [hunk 20]
```diff
@@ -0,0 +1,275 @@
...
+
+// todo:
+// consult with Guy whether to keep playlist and just update on disk or read/write...
+// I think its better to keep cache of playlist and just write to disk each update.
+
+class CrossDCsStreamPersistence {
+
+    constructor() {
+
+        this._presentationEndTimeRegex = /("presentationEndTime":\s*)(\d+)/;
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: this is generic and not per instance should belong ourside the class

File: lib/entry/crossDCsStreamPersistence.js [hunk 21]
```diff
@@ -0,0 +1,275 @@
...
+        // presentationEndTime value becomes past time before it is update by persistence loop.
+        this._timeBudget = 5000; // arbitrary time budget
+    }
+
+    streamPersistenceCheck(entryId, newEntry = false) {
+
+        let entryIsLive = false;
+        let ended = false;
+        let now = Date.now();
+        let nowStr = new Date(now).toLocaleString();
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why keep it as string?   dont' use presentationTimeStr and nowStr
also don't use LocaleString

File: lib/entry/crossDCsStreamPersistence.js [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+            })
+            .then((content) => {
+                let playlist = content;
+                this._playlistsData[entryId].lastProcessedTime = now;
+                if (newEntry) {
+                    logger.debug(`====> [streamPersistenceCheck] [${entryId}] STOPPED STREAMING expiration time set to [${nowStr}]`);
+                    playlist = playlist.replace(this._expireTimeRegex, `$1${now}`);
+                }
+                if (entryIsLive) {
+                    logger.debug(`====> [streamPersistenceCheck] [${entryId}] is LIVE in other DC, presentation end time set to [${presentationTimeStr}]`);
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: log line already prints the time at the beginning now need for extra string

File: lib/entry/crossDCsStreamPersistence.js [hunk 23]
```diff
@@ -0,0 +1,275 @@
...
+                }
+            });
+
+    }
+
+    flushEntryList() {
+
+        logger.debug(`====> [flushEntryList] updating persistence list on disk`);
+
+        if (Object.keys(this._entriesToProcess).length > 0) {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why all this code?
we can serialize the _entriesToProcess even if they are empty, this will simplify and remove lots of code

File: lib/entry/crossDCsStreamPersistence.js [hunk 24]
```diff
@@ -0,0 +1,275 @@
...
+            let entries2persistStr = JSON.stringify(_.keys(this._entriesToProcess));
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: please stringfy all  and not just the keys

File: lib/entry/crossDCsStreamPersistence.js [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+
+            return qio.write(this.stoppedEntriesFileName, entries2persistStr)
+                .then(() => {
+                    logger.debug(`====> [flushEntryList] Stopped entries: ${entries2persistStr}`);
+                })
+                .catch((err) => {
+                    if (err.code === 'ENOENT') {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: what is the scenario here?

File: lib/entry/crossDCsStreamPersistence.js [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+            delete this._entriesToProcess[entryId];
+            delete this._playlistsData[entryId];
+
+            /*        // todo: rethink flash operation location and timing...
+             return this.flushEntryList(this._entriesToProcess)
+             .catch((err) => {
+             logger.error(`failed to get entry server nodes for stopped entry [${entryId}], error: [${ErrorUtils.error2string(err)}]`);
+             return Q.resolve(true);
+             });*/
+            if (Object.keys(this._entriesToProcess).length === 0) {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: no need to check that, serialize anyways

File: lib/entry/crossDCsStreamPersistence.js [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+        }
+
+    }
+
+    removeEntry(entryId) {
+
+        logger.debug(`====> [removeEntry] removing [${entryId}] from persistence list`);
+        if (this._entriesToProcess.hasOwnProperty(entryId)) {
+            delete this._entriesToProcess[entryId];
+            delete this._playlistsData[entryId];
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: no need for _playlistsData

File: lib/entry/crossDCsStreamPersistence.js [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+            .then((exists) => {
+                if (exists) {
+                    return qio.read(this.stoppedEntriesFileName);
+                } else {
+                    return Q.reject(false);
+                }
+            })
+            .then((stoppedEntries) => {
+                let pAll = _.map(JSON.parse(stoppedEntries), (entryId) => {
+                    this.addEntry(entryId);
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: this this is not what you wanted

File: lib/entry/crossDCsStreamPersistence.js [hunk 29]
```diff
@@ -0,0 +1,275 @@
...
+                });
+                return Q.allSettled(pAll);
+            })
+            .catch((err) => {
+                if (_.isObject(err)) {
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: un-relaistic scneario

File: lib/entry/crossDCsStreamPersistence.js [hunk 30]
```diff
@@ -0,0 +1,275 @@
...
+                    if (err.code != 'ENOENT') {
+                        logger.info(`====> [init] error restoring stopped entries list, error: [${ErrorUtils.error2string(err)}]`);
+                        return Q.reject(err);
+                    } else {
+                        logger.warn(`${this.stoppedEntriesFileName} was not found`);
+                        return Q.resolve();
+                    }
+                } else {
+                    return qio.makeDirectory(stoppedEntriesPath)
```
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: better call qio.makeDirectory at the beginning and call finally (regardless if directoy exists or not)

**Comment by kalturaguy** [comment]
@lilachmaliniak  the code review is quite big, I recommend maybe you should look again on 0.15 branch and revisit your code
thanks

File: packager/config/nginx.conf.template (hunk 9 above)
Comment (lilachmaliniak) [inline packager/config/nginx.conf.template]: reverted.

File: lib/playlistGenerator/PlaylistGenerator.js (hunk 10 above)
Comment (lilachmaliniak) [inline lib/playlistGenerator/PlaylistGenerator.js]: yes. the business logic for the 2 values is implemented in crossDCsStreamPersistence.js, per your request the PlaylistGenrator is not involved.

File: lib/entry/crossDCsStreamPersistence.js (hunk 11 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: updated.

File: lib/entry/crossDCsStreamPersistence.js (hunk 12 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: Each time an entry is added I calculate the path and keep it in memory. It is never saved on disk. I don't see a problem with that.

File: lib/entry/crossDCsStreamPersistence.js (hunk 13 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: Init purpose is to restore all stopped entries from file when LiveController starts. Never the less I added log with entry id where relevant.

File: lib/entry/crossDCsStreamPersistence.js (hunk 14 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: changed

File: lib/playlistGenerator/PlaylistGenerator.js (hunk 10 above)
Comment (kalturaguy) [inline lib/playlistGenerator/PlaylistGenerator.js]: but you are calculating the same value not seprately, this is the bug

File: lib/entry/crossDCsStreamPersistence.js (hunk 12 above)
Comment (kalturaguy) [inline lib/entry/crossDCsStreamPersistence.js]: why do you keep somethign in memory when you can calcualte it every time you need it? please remove redudany data structures

File: lib/entry/crossDCsStreamPersistence.js (hunk 15 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: ok

Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: Let's discuss it.

File: common/config/config.json.template (hunk 17 above)
Comment (lilachmaliniak) [inline common/config/config.json.template]: I did merged from 1.0 on Thursday. this is not my code.

File: lib/Controller.js (hunk 18 above)
Comment (lilachmaliniak) [inline lib/Controller.js]: removed.

File: lib/Controller.js (hunk 19 above)
Comment (lilachmaliniak) [inline lib/Controller.js]: done.

File: lib/entry/crossDCsStreamPersistence.js (hunk 21 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: updated.

File: lib/entry/crossDCsStreamPersistence.js (hunk 23 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: removed.

File: lib/entry/crossDCsStreamPersistence.js (hunk 27 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: removed.

File: lib/entry/crossDCsStreamPersistence.js (hunk 30 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: done.

File: lib/entry/crossDCsStreamPersistence.js (hunk 26 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: removed

File: lib/entry/crossDCsStreamPersistence.js (hunk 25 above)
Comment (lilachmaliniak) [inline lib/entry/crossDCsStreamPersistence.js]: file not found from some reason.
